code/datums/material_container.dm

```diff
@@ -0,0 +1,250 @@
+		materials[MAT_BANANIUM] = new /datum/material/bananium()
+
+//For inserting an amount of material
+/datum/material_container/proc/insert_amount(amt, material_type = null)
+	if(amt > 0 && has_space(amt))
+		var/total_amount_saved = total_amount
+		if(material_type)
+			for(var/datum/material/M in materials)
+				if(M.material_type == material_type)
+					M.amount += amt
+					total_amount += amt
+		else
+			for(var/datum/material/M in materials)
+				M.amount += amt
+				total_amount += amt
+		return (total_amount - total_amount_saved)
+	return 0
+
+/datum/material_container/proc/insert_stack(obj/item/stack/S, amt = 0)
+	if(!amt)
+		amt = S.amount
+	var/material_amt = get_item_material_amount(S)
+	amt = min(amt, round(((max_amount - total_amount) / material_amt)))
+	if(!amt)
+		return 0
+
+	insert_materials(S,amt)
+	S.use(amt)
+	return amt
+
+/datum/material_container/proc/insert_item(obj/item/I)
+	if(!I)
+		return 0
+	if(istype(I,/obj/item/stack))
+		return insert_stack(I)
+
+	var/material_amount = get_item_material_amount(I)
+	if(!material_amount || !has_space(material_amount))
+		return 0
+
+	insert_materials(I)
+	return material_amount
+
+/datum/material_container/proc/insert_materials(obj/item/I, multiplier = 1) //for internal usage only
+	var/datum/material/M
+	for(var/MAT in materials)
+		M = materials[MAT]
+		M.amount += I.materials[MAT] * multiplier
+		total_amount += I.materials[MAT] * multiplier
+
+//For consuming material
+//mats is a list of types of material to use and the corresponding amounts, example: list(MAT_METAL=100, MAT_GLASS=200)
+/datum/material_container/proc/use_amount(list/mats)
+	if(!mats || !mats.len)
+		return 0
+
+	var/datum/material/M
+	for(var/MAT in materials)
+		M = materials[MAT]
+		if(M.amount < mats[MAT])
+			return 0
+
+	var/total_amount_save = total_amount
+	for(var/MAT in materials)
+		M = materials[MAT]
+		M.amount -= mats[MAT]
+		total_amount -= mats[MAT]
+
+	return total_amount_save - total_amount
+
+
+/datum/material_container/proc/use_amount_type(amt, material_type)
+	var/datum/material/M
+	M = materials[material_type]
+	if(M)
+		if(M.amount >= amt)
+			M.amount -= amt
+			total_amount -= amt
+			return amt
+	return 0
+
+//For spawning mineral sheets; internal use only
+/datum/material_container/proc/retrieve(sheet_amt, datum/material/M)
+	if(sheet_amt > 0 && M.amount >= (sheet_amt * MINERAL_MATERIAL_AMOUNT))
+		var/count = 0
+		var/obj/item/stack/sheet/S
+
+		while(sheet_amt > MAX_STACK_SIZE)
+			S = new M.sheet_type(get_turf(owner))
+			S.amount = MAX_STACK_SIZE
+			count += MAX_STACK_SIZE
+			M.amount -= MAX_STACK_SIZE * MINERAL_MATERIAL_AMOUNT
+			total_amount -= MAX_STACK_SIZE * MINERAL_MATERIAL_AMOUNT
+			sheet_amt -= MAX_STACK_SIZE
+
+		if(round(M.amount / MINERAL_MATERIAL_AMOUNT))
+			S = new M.sheet_type(get_turf(owner))
+			S.amount = sheet_amt
+			count += sheet_amt
+			M.amount -= sheet_amt * MINERAL_MATERIAL_AMOUNT
+			total_amount -= sheet_amt * MINERAL_MATERIAL_AMOUNT
```

Comment:
And here too could be used use_amount_type, you didnt made that proc for nothing right?